File: test_src/fortran/hierarchical_parallelism/atomic/target_parallel__for.f90

program target_parallel__for
    USE OMP_LIB

    implicit none
  
    INTEGER :: L = 5
    INTEGER :: i
    
    REAL :: COUNTER = 0

    
    

     
    
    !$OMP TARGET PARALLEL   MAP(TOFROM: COUNTER) 



    

    
    !$OMP DO 


    DO i = 1 , L 


    

    

!$OMP ATOMIC UPDATE

counter = counter + 1.


 
     

    END DO

    !$OMP END DO
     

    !$OMP END TARGET PARALLEL
    

    IF  ( ( ABS(COUNTER - L) ) > 10*EPSILON(COUNTER) ) THEN
        write(*,*)  'Expected L Got', COUNTER
        call exit(1)
    ENDIF

end program target_parallel__for

File: test_src/cpp/hierarchical_parallelism/atomic/float/target_parallel__for.cpp
#include <iostream>
#include <cmath>
#include <stdexcept>
#
bool almost_equal(float x, float gold, float tol) {
        return gold * (1-tol) <= x && x <= gold * ( 1+tol );
}
void test_target_parallel__for(){
 // Input and Outputs
 const int L = 262144;
float counter{};
// Main program
#pragma omp target parallel  map(tofrom:counter) 
{
#pragma omp for 
    for (int i = 0 ; i < L ; i++ )
{
#pragma omp atomic update
counter += float { 1 };
    } 
    } 
// Validation
if ( !almost_equal(counter,float { L }, 0.1)  ) {
    std::cerr << "Expected: " << L << " Got: " << counter << std::endl;
    throw std::runtime_error( "target_parallel__for give incorect value when offloaded");
}
}
int main()
{
    test_target_parallel__for();
}
